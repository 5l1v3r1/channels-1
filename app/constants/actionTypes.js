// @flow
const actionTypes = {
  REQUEST_ACCESS_TOKEN: 'REQUEST_ACCESS_TOKEN',
  RECEIVE_ACCESS_TOKEN: 'RECEIVE_ACCESS_TOKEN',
  ACCESS_TOKEN_ERROR: 'ACCESS_TOKEN_ERROR'
};

export default actionTypes;
