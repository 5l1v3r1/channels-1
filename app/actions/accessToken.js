// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
