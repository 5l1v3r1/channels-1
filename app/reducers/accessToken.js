// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

function token(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
      return payload;
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_REQUEST:
      return true;
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
    case actionTypes.FETCH_ACCESS_TOKEN_FAILURE:
      return false;
    default:
      return state;
  }
}


function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_REQUEST:
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
      return false;
    case actionTypes.FETCH_ACCESS_TOKEN_FAILURE:
      return true;
    default:
      return state;
  }
}

const accessToken = combineReducers({
  token,
  isFetching,
  error
});

export default accessToken;
