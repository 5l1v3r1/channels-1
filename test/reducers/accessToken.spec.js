import accessToken from '../../app/reducers/accessToken';
import actionTypes from '../../app/constants/actionTypes';

describe('accessToken reducer', () => {
  it('should return the initial state', () => {
    expect(accessToken(undefined, {})).toMatchSnapshot();
  });

  it('should handle FETCH_ACCESS_TOKEN_REQUEST', () => {
    const action = { type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST };
    expect(accessToken(undefined, action)).toMatchSnapshot();
  });

  it('should handle FETCH_ACCESS_TOKEN_SUCCESS', () => {
    const state = {
      token: '',
      isFetching: true,
      error: false
    };
    const action = {
      type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
      payload: 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg'
    };
    expect(accessToken(state, action)).toMatchSnapshot();
  });

  it('should handle FETCH_ACCESS_TOKEN_FAILURE', () => {
    const state = {
      token: '',
      isFetching: true,
      error: false
    };
    const action = { type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE };
    expect(accessToken(state, action)).toMatchSnapshot();
  });

  it('should handle unknown action type', () => {
    const action = { type: 'unknown' };
    expect(accessToken(undefined, action)).toMatchSnapshot();
  });
});
