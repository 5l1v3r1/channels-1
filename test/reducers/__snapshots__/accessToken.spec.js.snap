// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessToken reducer should handle ACCESS_TOKEN_ERROR 1`] = `
Object {
  "accessToken": "",
  "error": true,
  "isFetching": false,
}
`;

exports[`accessToken reducer should handle RECEIVE_ACCESS_TOKEN 1`] = `
Object {
  "accessToken": "1/BYafNRmup5QmMT8NmWJemosgtOF4GCu8-sqMIO7nHBk",
  "error": false,
  "isFetching": false,
}
`;

exports[`accessToken reducer should handle REQUEST_ACCESS_TOKEN 1`] = `
Object {
  "accessToken": "",
  "error": false,
  "isFetching": true,
}
`;

exports[`accessToken reducer should handle unknown action type 1`] = `
Object {
  "accessToken": "",
  "error": false,
  "isFetching": false,
}
`;

exports[`accessToken reducer should return the initial state 1`] = `
Object {
  "accessToken": "",
  "error": false,
  "isFetching": false,
}
`;
